### ISSUE TITLE
Invalid position reported after scanStringLiteral hits bad code

### ISSUE THREAD

**Issue by Arcanemagus** [jquery/esprima#1457]
It appears that in the `scanStringLiteral()` function [here](https://github.com/jquery/esprima/blob/2.7.1/esprima.js#L1056) when scanning an invalid "string" that doesn't finish before the end of the line, the position reported as the invalid character is incorrect.

What it looks like is going on is when it hits the line ending and `break;`s [here](https://github.com/jquery/esprima/blob/2.7.1/esprima.js#L1132) the current index will have been incremented past the end of the line [here](https://github.com/jquery/esprima/blob/2.7.1/esprima.js#L1067) and is never corrected since it has broken out of the loop and is in the process of calling `throwUnexpectedToken()`

When linting this code:

```JavaScript
var testError = image/png'
```

This results in incorrect results when a linter is using `esprima` for parsing:
![image](https://cloud.githubusercontent.com/assets/427137/12569661/b88c5760-c385-11e5-9bee-a4c44c983df2.png)

Originally filed over [here](https://github.com/jscs-dev/node-jscs/issues/2083), with reports that seem to be related to this coming in from users of `jscs` and `jshint` from the Atom linter plugins.

Version: `esprima` v2.7.1
Operating System: Windows 10 x64

If you need any further information I'd be happy to help.

**Comment by ariya** [comment]
Thank you for the detailed analysis!

This problem can be reproduced with this quick test case:
```js
> try { require('esprima').parse('x = y"') } catch (e) { console.log(e) }
{ [Error: Line 1: Unexpected token ILLEGAL]
  lineNumber: 1,
  description: 'Unexpected token ILLEGAL',
  index: 6 }
```
In the above output, `index` (zero based) should not be really greater than the line length.

**Comment by ariya** [comment]
I am working on a fix and I will keep everyone updated.

**Comment by ariya** [comment]
I will also backport the fix to 2.7.

**Comment by Arcanemagus** [comment]
Thanks for fixing this!

Now we just have to wait on the linters to incorporate the update, and then users to update their systems and I'll finally stop getting bug reports, but such is the way things work. :stuck_out_tongue:

**Comment by ariya** [comment]
@Arcanemagus Sounds good, I'm trying to get 2.7.2 out this evening.

**Comment by ariya** [comment]
I've published esprima@2.7.2 on npm that contains this fix.
